plugins/lookup/etcd3.py

```diff
@@ -0,0 +1,230 @@
+        type: list
+        elements: complex
+'''
+
+import os
+import re
+import traceback
+
+from ansible.errors import AnsibleLookupError
+from ansible.utils.display import Display
+from ansible.plugins.lookup import LookupBase
+from ansible.module_utils.basic import missing_required_lib
+from ansible.module_utils._text import to_native
+
+try:
+    import etcd3
+    HAS_ETCD = True
+except ImportError:
+    HAS_ETCD = False
+
+display = Display()
+
+
+def etcd3_client(client_params):
+    etcd = None
+    try:
+        etcd = etcd3.client(**client_params)
+        etcd.status()
+    except Exception as exp:
+        raise AnsibleLookupError('Cannot connect to etcd cluster: %s' % (to_native(exp)))
+    return etcd
+
+
+class LookupModule(LookupBase):
+
+    def log_env_override(self, params, key):
+        display.verbose(
+            "Overriding etcd3 '{0}' param with ENV variable '{1}'".format(
+                to_native(key),
+                to_native(params[key])
+            )
+        )
+
+    def run(self, terms, variables, **kwargs):
+
+        self.set_options(var_options=variables, direct=kwargs)
+
+        if not HAS_ETCD:
+            display.error(missing_required_lib('etcd3'))
+            return None
+
+        # get etcd3 optional args
+        client_params = {
+            'host': '127.0.0.1',
+            'port': 2379,
+            'ca_cert': None,
+            'cert_cert': None,
+            'cert_key': None,
+            'timeout': 60,
+            'user': None,
+            'password': None
+        }
+
+        # use ETCD environment variables if they are set
+        etcd3_envs = {
+            'ca_cert': 'ETCDCTL_CACERT',
+            'cert_cert': 'ETCDCTL_CERT',
+            'cert_key': 'ETCDCTL_KEY',
+            'user': 'ETCDCTL_USER',
+            'password': 'ETCDCTL_PASSWORD',
+            'timeout': 'ETCDCTL_DIAL_TIMEOUT',
+        }
+
+        for key in etcd3_envs.keys():
+            if etcd3_envs[key] in os.environ:
+                client_params[key] = os.environ[etcd3_envs[key]]
+                self.log_env_override(client_params, key)
+
+        # ETCDCTL_ENDPOINTS expects something in the form <http(s)://server:port> of <server:port>
+        # so here we use a regex to extract server and port
+        if 'ETCDCTL_ENDPOINTS' in os.environ:
+            match = re.compile(
+                r'^(https?://)?(?P<host>(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})|([-_\d\w\.]+))(:(?P<port>\d{1,5}))?/?$'
+            ).match(os.environ['ETCDCTL_ENDPOINTS'])
+            if match:
+                if match.group('host'):
+                    client_params['host'] = match.group('host')
+                    self.log_env_override(client_params, 'host')
+                if match.group('port'):
+                    client_params['port'] = match.group('port')
+                    self.log_env_override(client_params, 'port')
+
+        # override env with optional lookup parameters
+        for key in client_params.keys():
```

Comment:
You should really use `get_option()`. Check out how the `etcd` lookup is doing this.

Reply:
Hi Felix, look great, code is smarter, and it works as expected... except that it breaks unit tests !!!
I get the following error:
```
________________________________________________________ TestLookupModule.test_key _________________________________________________________
[gw8] linux2 -- Python 2.7.16 /tmp/python-r5IC1X-ansible/python
self = <ansible_collections.community.general.tests.unit.plugins.lookup.test_etcd3.TestLookupModule testMethod=test_key>

    @patch('ansible_collections.community.general.plugins.lookup.etcd3.etcd3_client', FakeEtcd3Client())
    def test_key(self):
        expected_result = [{'key': 'a_key', 'value': 'a_key value'}]
        lookup = etcd3.LookupModule()
        lookup._load_name = 'etcd3'
>       self.assertListEqual(expected_result, lookup.run(['a_key'], []))

tests/unit/plugins/lookup/test_etcd3.py:51: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
plugins/lookup/etcd3.py:162: in run
    if self.get_option('endpoints'):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ansible_collections.community.general.plugins.lookup.etcd3.LookupModule object at 0x7fb9a0884f50>
option = 'endpoints', hostvars = None

    def get_option(self, option, hostvars=None):
        if option not in self._options:
            try:
                option_value = C.config.get_config_value(option, plugin_type=get_plugin_class(self), plugin_name=self._load_name, variables=hostvars)
            except AnsibleError as e:
>               raise KeyError(to_native(e))
E               KeyError: 'Requested entry (plugin_type: lookup plugin: etcd3 setting: endpoints ) was not defined in configuration.'

../../../ansible/lib/ansible/plugins/__init__.py:60: KeyError
```

I build my test class on `tests/unit/plugins/lookup/test_etcd3.py` that way:

```python
class TestLookupModule(unittest.TestCase):

    def setUp(self):
        etcd3.HAS_ETCD = True
        self.lookup = etcd3.LookupModule()
        self.lookup._load_name = 'etcd3'

    @patch('ansible_collections.community.general.plugins.lookup.etcd3.etcd3_client', FakeEtcd3Client())
    def test_key(self):
        expected_result = [{'key': 'a_key', 'value': 'a_key value'}]
        self.assertListEqual(expected_result, self.lookup.run(['a_key'], []))
```

It looks like in that context, `contstants` fails to load configuration from etcd3.py source file. I didn't understood all the magic behind so I'm bit stuck and I can't figure _why_ this doesn't work.

It tried to mangle the LookupModule `_load_name` class attribute with 'etcd3' and 'community.general.etcd3' wioth no effect.

Any idea ?

Reply:
Forgot to spécify my updated DOCUMENTATION docstring:
```yaml
    author:
    - Eric Belhomme <ebelhomme@fr.scc.com>
    lookup: community.general.etcd3
    short_description: Get key values from etcd3 server
    description:
    - Retrieves key values and/or key prefixes from etcd3 using gRPC API.

    options:
        _terms:
            description:
            - The list of keys to look up on the etcd3 server.
            type: list
            elements: str
            required: True
        prefix:
            description:
            - Look for key or prefix key.
            type: bool
        host:
            description:
            - etcd3 listening client host.
            default: '127.0.0.1'
            type: str
        port:
            description:
            - etcd3 listening client port.
            default: 2379
            type: int
        ca_cert:
            description:
            - etcd3 CA authority.
            env:
            - name: ETCDCTL_CACERT
            type: str
        cert_cert:
            description:
            - etcd3 client certificate.
            env:
            - name: ETCDCTL_CERT
            type: str
        cert_key:
            description:
            - etcd3 client private key.
            env:
            - name: ETCDCTL_KEY
            type: str
        timeout:
            description:
            - Client timeout.
            default: 60
            env:
            - name: ETCDCTL_DIAL_TIMEOUT
            type: int
        user:
            description:
            - Authentified user name.
            env:
            - name: ETCDCTL_USER
            type: str
        password:
            description:
            - Authentified user password.
            env:
            - name: ETCDCTL_PASSWORD
            type: str
        endpoints:
            description:
            - Counterpart of ETCDCTL_ENDPOINTS enviroment variable.
              Specify the etcd3 connection with and URL form or <host>:<port> form
            env:
            - name: ETCDCTL_ENDPOINTS
            type: str

    requirements:
    - "etcd3 >= 0.10"
```

I tried to add `default: None` and `required: True` to the `endpoints` dict, but with no effect

Reply:
It would be nice if you could push your latest code, so that we can try it out.

Reply:
Done. I also rebased my branch from upstream's master

Reply:
Interestingly, there are no lookup unit tests for lookup plugins which use `get_options` (at least neither in ansible/ansible nor in community.general). (Except the `conjur_variable` lookup test, but that doesn't actually run the lookup itself, it only tests functions from that lookup.)

So maybe we're writing history here ;-)

Reply:
I was able to get it to run by using the proper lookup loader:
```.diff
diff --git a/tests/unit/plugins/lookup/test_etcd3.py b/tests/unit/plugins/lookup/test_etcd3.py
index d9b86c9..cb33010 100644
--- a/tests/unit/plugins/lookup/test_etcd3.py
+++ b/tests/unit/plugins/lookup/test_etcd3.py
@@ -13,6 +13,8 @@ from ansible_collections.community.general.tests.unit.compat.mock import patch,
 from ansible.errors import AnsibleError
 from ansible_collections.community.general.plugins.lookup import etcd3
 
+from ansible.plugins.loader import lookup_loader
+
 
 class FakeKVMetadata:
 
@@ -39,8 +41,7 @@ class TestLookupModule(unittest.TestCase):
 
     def setUp(self):
         etcd3.HAS_ETCD = True
-        self.lookup = etcd3.LookupModule()
-        self.lookup._load_name = 'community.general.etcd3'
+        self.lookup = lookup_loader.get('community.general.etcd3')
 
     @patch('ansible_collections.community.general.plugins.lookup.etcd3.etcd3_client', FakeEtcd3Client())
     def test_key(self):
```
(The [tests](https://github.com/ansible/ansible/blob/devel/test/units/plugins/lookup/test_env.py#L21) of core's `env` plugin used `lookup_loader`, so I decided to try that... and it works! This is probably how Ansible loads lookup plugins, so that should do all necessary set-up.)

Reply:
you rocks !
thanks for the patch :)